### PR TITLE
update nullability of drawAtlas methods and flesh out docs

### PR DESCRIPTION
**Pull request by flar**
These changes fix two outstanding issues with the `Canvas.drawAtlas` and `Canvas.drawRawAtlas` methods:

- https://github.com/flutter/flutter/issues/62457 (unnecessary non-null restrictions on their parameters)
- https://github.com/flutter/flutter/issues/62450 (document their operation in more detail)

The changes bring the null-ness of 2 of their parameters into compliance with the Skia implementations and the needs of common uses of these methods. Since nulls will now be allowed where they were not before, the changes represent a relaxing of restrictions and so should be backwards compatible with existing apps that use these methods.

The implementation is complete for the regular engine and includes new tests, but the web_ui version of these classes will need some updates to match (and I expect some check failures on this initial push due to the temporarily mismatched method signatures).

The changes are ready to have their API changes reviewed in their current form, though.

### CODE REVIEW

**Comment by flar** [comment]
This is odd, I haven't updated the web_ui signatures, but I guess the API tests don't test the nullability parts of the signatures.

I still have to update the implementation of the methods in web_ui to match the changes in the lib/ui files, though.

**Comment by flar** [comment]
The web_ui files are all updated as far as I could understand them. I pasted in the new lengthy doc comments even though I understand that the doc comments in the web_ui api file are not actually used to generate any documents...?

File: lib/ui/painting.dart [hunk 1]
```diff
@@ -4025,13 +4025,117 @@ class Canvas extends NativeFieldWrapperClass2 {
...
+  /// be either null or an empty list and the [blendMode] argument may also be null.
+  ///
+  /// The optional [cullRect] argument can provide an estimate of the bounds of the
+  /// coordinates rendered by all components of the atlas to be compared against
+  /// the clip to quickly reject the operation if it does not intersect.
+  ///
+  /// An example usage to render many sprites from a single sprite atlas with no
+  /// rotations or scales:
+  /// ```dart
+  /// class Sprite { int index; double centerX; double centerY; }
```
Comment (dnfield) [inline lib/ui/painting.dart:4066]: Not: format this on separate lines

File: lib/ui/painting.dart [hunk 2]
```diff
@@ -4025,13 +4025,117 @@ class Canvas extends NativeFieldWrapperClass2 {
...
+  ///
+  ///   ...
+  /// }
+  /// ```
   ///
-  /// This method allows for optimization when you only want to draw part of an
-  /// image on the canvas, such as when using sprites or zooming. It is more
-  /// efficient than using clips or masks directly.
+  /// Another example usage which renders sprites with an optional opacity and rotation:
+  /// ```dart
```
Comment (dnfield) [inline lib/ui/painting.dart:4099]: Nit: new line between blocks

File: lib/ui/painting.dart [hunk 3]
```diff
@@ -4025,13 +4025,117 @@ class Canvas extends NativeFieldWrapperClass2 {
...
+  ///         Color.white.withAlpha(sprite.alpha),
+  ///     ], BlendMode.srcIn, null, paint);
+  ///   }
+  ///
+  ///   ...
+  /// }
+  /// ```
   ///
-  /// All parameters must not be null.
+  /// The [atlas], [transforms], [rects], and [paint] arguments must not be null.
```
Comment (dnfield) [inline lib/ui/painting.dart:4135]: I think we can remove this line with nnbd now

File: lib/ui/painting.dart [hunk 4]
```diff
@@ -4102,7 +4212,111 @@ class Canvas extends NativeFieldWrapperClass2 {
...
+  ///     Paint paint = Paint();
+  ///     canvas.drawAtlas(spriteAtlas, transformList, rectList, null, null, null, paint);
+  ///   }
+  ///
+  ///   ...
+  /// }
+  /// ```
+  ///
+  /// Another example usage which renders sprites with an optional opacity and rotation:
+  /// ```dart
```
Comment (dnfield) [inline lib/ui/painting.dart:4261]: Ditto

File: lib/ui/painting.dart (hunk 1 above)
Comment (flar) [reply lib/ui/painting.dart:4066]: Done.

**Comment by dnfield** [comment]
I haven't double checked but isn't the C++ side expecting these to not be Dart_Null?  I really like the doc updates but it's not clear to me why we would start accepting null instead of an empty array.  I think in general it makes the code more complex and if you forget the null check somewhere you crash.

File: lib/ui/painting.dart (hunk 2 above)
Comment (flar) [reply lib/ui/painting.dart:4099]: I'm not sure what you are asking for here. There is a newline at 4097, or are you asking for another one after 4098?

Comment (flar) [reply lib/ui/painting.dart:4099]: Checking other uses I understand now. Done (actually doing).

File: lib/ui/painting.dart (hunk 3 above)
Comment (flar) [reply lib/ui/painting.dart:4135]: OK

**Comment by flar** [comment]
> I haven't double checked but isn't the C++ side expecting these to not be Dart_Null? I really like the doc updates but it's not clear to me why we would start accepting null instead of an empty array. I think in general it makes the code more complex and if you forget the null check somewhere you crash.

I looked through and it appears that they handle null lists being sent down in the arg converters. I am certainly running a test case where I send down nulls and it behaves as expected.

**Comment by flar** [comment]
Tagging @chinmaygarde regarding the question about whether the nulls will be handled correctly by the DartNative argument converters. Since an empty colors array used to translate into a null value being sent down, I think this is mostly a moot question, but it would be good to verify that my null handling here is kosher with respect to the native code.

**Comment by jason-simmons** [comment]
The Tonic wrapper for typed lists such as `Int32List` will construct an empty list if the converted Dart handle is a Dart null.

See https://github.com/flutter/engine/blob/master/third_party/tonic/typed_data/typed_list.cc#L20

**Comment by flar** [comment]
> The Tonic wrapper for typed lists such as `Int32List` will construct an empty list if the converted Dart handle is a Dart null.
> 
> See https://github.com/flutter/engine/blob/master/third_party/tonic/typed_data/typed_list.cc#L20

It looks like the `data_` field, which is what the value of the argument is, will be `nullptr` in that case.

See https://github.com/flutter/engine/blob/master/third_party/tonic/typed_data/typed_list.cc#L19
and https://github.com/flutter/engine/blob/master/third_party/tonic/typed_data/typed_list.h#L40
and https://github.com/flutter/engine/blob/master/lib/ui/painting/canvas.cc#L462

**Comment by flar** [comment]
Does this need to go through any formal API review?

**Comment by dnfield** [comment]
Nah. If you're concerned though feel free to ping @hixie or @zanderso

**Pull request merged**